File: pdu/python/ev3_msgs/pdu_conv_Ev3PduSensor.py

import struct
from .pdu_pytype_Ev3PduSensor import Ev3PduSensor
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class
from ..ev3_msgs.pdu_conv_Ev3PduSensorHeader import *
from ..ev3_msgs.pdu_conv_Ev3PduColorSensor import *
from ..ev3_msgs.pdu_conv_Ev3PduTouchSensor import *



def pdu_to_py_Ev3PduSensor(binary_data: bytearray) -> Ev3PduSensor:
    py_obj = Ev3PduSensor()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_Ev3PduSensor(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_Ev3PduSensor(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: Ev3PduSensor, base_off: int):
    # array_type: single 
    # data_type: struct 
    # member_name: head 
    # type_name: Ev3PduSensorHeader 
    # offset: 0 size: 152 
    # array_len: 1

    tmp_py_obj = Ev3PduSensorHeader()
    binary_read_recursive_Ev3PduSensorHeader(meta, binary_data, tmp_py_obj, base_off + 0)
    py_obj.head = tmp_py_obj
    
    # array_type: array 
    # data_type: primitive 
    # member_name: buttons 
    # type_name: uint8 
    # offset: 152 size: 1 
    # array_len: 1

    
    array_value = binary_io.readBinary(binary_data, base_off + 152, 1)
    py_obj.buttons = binary_io.binToArrayValues("uint8", array_value)
    
    # array_type: array 
    # data_type: struct 
    # member_name: color_sensors 
    # type_name: Ev3PduColorSensor 
    # offset: 156 size: 40 
    # array_len: 2

    i = 0
    array_size = 2
    one_elm_size = int(40 / array_size)
    array_value = []
    while i < array_size:
        tmp_py_obj = Ev3PduColorSensor()
        binary_read_recursive_Ev3PduColorSensor(meta, binary_data, tmp_py_obj, base_off + 156 + (i * one_elm_size))
        array_value.append(tmp_py_obj)
        i = i + 1
    py_obj.color_sensors = array_value    
    
    # array_type: array 
    # data_type: struct 
    # member_name: touch_sensors 
    # type_name: Ev3PduTouchSensor 
    # offset: 196 size: 8 
    # array_len: 2

    i = 0
    array_size = 2
    one_elm_size = int(8 / array_size)
    array_value = []
    while i < array_size:
        tmp_py_obj = Ev3PduTouchSensor()
        binary_read_recursive_Ev3PduTouchSensor(meta, binary_data, tmp_py_obj, base_off + 196 + (i * one_elm_size))
        array_value.append(tmp_py_obj)
        i = i + 1
    py_obj.touch_sensors = array_value    
    
    # array_type: array 
    # data_type: primitive 
    # member_name: motor_angle 
    # type_name: uint32 
    # offset: 204 size: 12 
    # array_len: 3

    
    array_value = binary_io.readBinary(binary_data, base_off + 204, 12)
    py_obj.motor_angle = binary_io.binToArrayValues("uint32", array_value)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: gyro_degree 
    # type_name: int32 
    # offset: 216 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 216, 4)
    py_obj.gyro_degree = binary_io.binTovalue("int32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: gyro_degree_rate 
    # type_name: int32 
    # offset: 220 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 220, 4)
    py_obj.gyro_degree_rate = binary_io.binTovalue("int32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: sensor_ultrasonic 
    # type_name: uint32 
    # offset: 224 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 224, 4)
    py_obj.sensor_ultrasonic = binary_io.binTovalue("uint32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: gps_lat 
    # type_name: float64 
    # offset: 232 size: 8 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 232, 8)
    py_obj.gps_lat = binary_io.binTovalue("float64", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: gps_lon 
    # type_name: float64 
    # offset: 240 size: 8 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 240, 8)
    py_obj.gps_lon = binary_io.binTovalue("float64", bin)
    
    return py_obj


def py_to_pdu_Ev3PduSensor(py_obj: Ev3PduSensor) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_Ev3PduSensor(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_Ev3PduSensor(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: Ev3PduSensor):
    # array_type: single 
    # data_type: struct 
    # member_name: head 
    # type_name: Ev3PduSensorHeader 
    # offset: 0 size: 152 
    # array_len: 1
    type = "Ev3PduSensorHeader"
    off = 0

    binary_write_recursive_Ev3PduSensorHeader(parent_off + off, bw_container, allocator, py_obj.head)
    
    # array_type: array 
    # data_type: primitive 
    # member_name: buttons 
    # type_name: uint8 
    # offset: 152 size: 1 
    # array_len: 1
    type = "uint8"
    off = 152

    
    elm_size =  1 
    array_size = int(1.0)
    one_elm_size = int(elm_size / array_size)
    binary = binary_io.typeTobin_array(type, py_obj.buttons, one_elm_size)
    allocator.add(binary, expected_offset=(parent_off + off))
    
    # array_type: array 
    # data_type: struct 
    # member_name: color_sensors 
    # type_name: Ev3PduColorSensor 
    # offset: 156 size: 40 
    # array_len: 2
    type = "Ev3PduColorSensor"
    off = 156

    for i, elm in enumerate(py_obj.color_sensors):
        elm_size = 40
        array_size = int(20.0)
        one_elm_size = int(elm_size / array_size)
        binary_write_recursive_Ev3PduColorSensor((parent_off + off + i * one_elm_size), bw_container, allocator, elm)
    
    # array_type: array 
    # data_type: struct 
    # member_name: touch_sensors 
    # type_name: Ev3PduTouchSensor 
    # offset: 196 size: 8 
    # array_len: 2
    type = "Ev3PduTouchSensor"
    off = 196

    for i, elm in enumerate(py_obj.touch_sensors):
        elm_size = 8
        array_size = int(4.0)
        one_elm_size = int(elm_size / array_size)
        binary_write_recursive_Ev3PduTouchSensor((parent_off + off + i * one_elm_size), bw_container, allocator, elm)
    
    # array_type: array 
    # data_type: primitive 
    # member_name: motor_angle 
    # type_name: uint32 
    # offset: 204 size: 12 
    # array_len: 3
    type = "uint32"
    off = 204

    
    elm_size =  12 
    array_size = int(4.0)
    one_elm_size = int(elm_size / array_size)
    binary = binary_io.typeTobin_array(type, py_obj.motor_angle, one_elm_size)
    allocator.add(binary, expected_offset=(parent_off + off))
    
    # array_type: single 
    # data_type: primitive 
    # member_name: gyro_degree 
    # type_name: int32 
    # offset: 216 size: 4 
    # array_len: 1
    type = "int32"
    off = 216

    
    bin = binary_io.typeTobin(type, py_obj.gyro_degree)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: gyro_degree_rate 
    # type_name: int32 
    # offset: 220 size: 4 
    # array_len: 1
    type = "int32"
    off = 220

    
    bin = binary_io.typeTobin(type, py_obj.gyro_degree_rate)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: sensor_ultrasonic 
    # type_name: uint32 
    # offset: 224 size: 4 
    # array_len: 1
    type = "uint32"
    off = 224

    
    bin = binary_io.typeTobin(type, py_obj.sensor_ultrasonic)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: gps_lat 
    # type_name: float64 
    # offset: 232 size: 8 
    # array_len: 1
    type = "float64"
    off = 232

    
    bin = binary_io.typeTobin(type, py_obj.gps_lat)
    bin = get_binary(type, bin, 8)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: gps_lon 
    # type_name: float64 
    # offset: 240 size: 8 
    # array_len: 1
    type = "float64"
    off = 240

    
    bin = binary_io.typeTobin(type, py_obj.gps_lon)
    bin = get_binary(type, bin, 8)
    allocator.add(bin, expected_offset=parent_off + off)
    

if __name__ == "__main__":
    import sys
    import json

    def print_usage():
        print(f"Usage: python -m pdu.python.pdu_conv_Ev3PduSensor <read|write> [args...]")
        print(f"  read <input_binary_file> <output_json_file>")
        print(f"  write <input_json_file> <output_binary_file>")

    if len(sys.argv) < 2:
        print_usage()
        sys.exit(1)

    command = sys.argv[1]

    if command == "read":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)
        
        binary_filepath = sys.argv[2]
        output_json_filepath = sys.argv[3]

        with open(binary_filepath, "rb") as f:
            binary_data = bytearray(f.read())
        
        py_obj = pdu_to_py_Ev3PduSensor(binary_data)
        
        with open(output_json_filepath, "w") as f:
            f.write(py_obj.to_json())

    elif command == "write":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)

        input_json_filepath = sys.argv[2]
        output_binary_filepath = sys.argv[3]

        with open(input_json_filepath, "r") as f:
            json_str = f.read()
        
        py_obj = Ev3PduSensor.from_json(json_str)
        
        binary_data = py_to_pdu_Ev3PduSensor(py_obj)

        with open(output_binary_filepath, "wb") as f:
            f.write(binary_data)

    else:
        print(f"Unknown command: {command}")
        print_usage()
        sys.exit(1)

File: pdu/javascript/ev3_msgs/pdu_conv_Ev3PduSensor.js
import * as PduUtils from '../pdu_utils.js';
import { Ev3PduSensor } from './pdu_jstype_Ev3PduSensor.js';
import { Ev3PduSensorHeader } from './pdu_jstype_Ev3PduSensorHeader.js';
import { binary_read_recursive_Ev3PduSensorHeader, binary_write_recursive_Ev3PduSensorHeader } from './pdu_conv_Ev3PduSensorHeader.js';
import { Ev3PduColorSensor } from './pdu_jstype_Ev3PduColorSensor.js';
import { binary_read_recursive_Ev3PduColorSensor, binary_write_recursive_Ev3PduColorSensor } from './pdu_conv_Ev3PduColorSensor.js';
import { Ev3PduTouchSensor } from './pdu_jstype_Ev3PduTouchSensor.js';
import { binary_read_recursive_Ev3PduTouchSensor, binary_write_recursive_Ev3PduTouchSensor } from './pdu_conv_Ev3PduTouchSensor.js';


/**
 * Deserializes a binary PDU into a Ev3PduSensor object.
 * @param {ArrayBuffer} binary_data
 * @returns { Ev3PduSensor }
 */
export function pduToJs_Ev3PduSensor(binary_data) {
    const js_obj = new Ev3PduSensor();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_Ev3PduSensor(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_Ev3PduSensor(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: head, type: Ev3PduSensorHeader (struct)

    {
        const tmp_obj = new Ev3PduSensorHeader();
        binary_read_recursive_Ev3PduSensorHeader(meta, binary_data, tmp_obj, base_off + 0);
        js_obj.head = tmp_obj;
    }
    
    // member: buttons, type: uint8 (primitive)

    
    {
        const array_bin = PduUtils.readBinary(binary_data, base_off + 152, 1);
        js_obj.buttons = PduUtils.binToArrayValues("uint8", array_bin, 1);
    }
    
    // member: color_sensors, type: Ev3PduColorSensor (struct)

    {
        const array_size = 2;
        const one_elm_size = 40 / array_size;
        const array_value = [];
        for (let i = 0; i < array_size; i++) {
            const tmp_obj = new Ev3PduColorSensor();
            const item_offset = base_off + 156 + (i * one_elm_size);
            binary_read_recursive_Ev3PduColorSensor(meta, binary_data, tmp_obj, item_offset);
            array_value.push(tmp_obj);
        }
        js_obj.color_sensors = array_value;
    }
    
    // member: touch_sensors, type: Ev3PduTouchSensor (struct)

    {
        const array_size = 2;
        const one_elm_size = 8 / array_size;
        const array_value = [];
        for (let i = 0; i < array_size; i++) {
            const tmp_obj = new Ev3PduTouchSensor();
            const item_offset = base_off + 196 + (i * one_elm_size);
            binary_read_recursive_Ev3PduTouchSensor(meta, binary_data, tmp_obj, item_offset);
            array_value.push(tmp_obj);
        }
        js_obj.touch_sensors = array_value;
    }
    
    // member: motor_angle, type: uint32 (primitive)

    
    {
        const array_bin = PduUtils.readBinary(binary_data, base_off + 204, 12);
        js_obj.motor_angle = PduUtils.binToArrayValues("uint32", array_bin, 3);
    }
    
    // member: gyro_degree, type: int32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 216, 4);
        js_obj.gyro_degree = PduUtils.binToValue("int32", bin);
    }
    
    // member: gyro_degree_rate, type: int32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 220, 4);
        js_obj.gyro_degree_rate = PduUtils.binToValue("int32", bin);
    }
    
    // member: sensor_ultrasonic, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 224, 4);
        js_obj.sensor_ultrasonic = PduUtils.binToValue("uint32", bin);
    }
    
    // member: gps_lat, type: float64 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 232, 8);
        js_obj.gps_lat = PduUtils.binToValue("float64", bin);
    }
    
    // member: gps_lon, type: float64 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 240, 8);
        js_obj.gps_lon = PduUtils.binToValue("float64", bin);
    }
    
    return js_obj;
}

/**
 * Serializes a Ev3PduSensor object into a binary PDU.
 * @param { Ev3PduSensor } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_Ev3PduSensor(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_Ev3PduSensor(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_Ev3PduSensor(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: head, type: Ev3PduSensorHeader (struct)

    {
        binary_write_recursive_Ev3PduSensorHeader(parent_off + 0, bw_container, allocator, js_obj.head);
    }
    
    // member: buttons, type: uint8 (primitive)

    
    {
        const buffer = PduUtils.typesToBin("uint8", js_obj.buttons);
        allocator.add(buffer, parent_off + 152);
    }
    
    // member: color_sensors, type: Ev3PduColorSensor (struct)

    {
        const one_elm_size = 40 / 2;
        for (let i = 0; i < 2; i++) {
            binary_write_recursive_Ev3PduColorSensor(parent_off + 156 + (i * one_elm_size), bw_container, allocator, js_obj.color_sensors[i]);
        }
    }
    
    // member: touch_sensors, type: Ev3PduTouchSensor (struct)

    {
        const one_elm_size = 8 / 2;
        for (let i = 0; i < 2; i++) {
            binary_write_recursive_Ev3PduTouchSensor(parent_off + 196 + (i * one_elm_size), bw_container, allocator, js_obj.touch_sensors[i]);
        }
    }
    
    // member: motor_angle, type: uint32 (primitive)

    
    {
        const buffer = PduUtils.typesToBin("uint32", js_obj.motor_angle);
        allocator.add(buffer, parent_off + 204);
    }
    
    // member: gyro_degree, type: int32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int32", js_obj.gyro_degree, 4);
        allocator.add(bin, parent_off + 216);
    }
    
    // member: gyro_degree_rate, type: int32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("int32", js_obj.gyro_degree_rate, 4);
        allocator.add(bin, parent_off + 220);
    }
    
    // member: sensor_ultrasonic, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.sensor_ultrasonic, 4);
        allocator.add(bin, parent_off + 224);
    }
    
    // member: gps_lat, type: float64 (primitive)

    
    {
        const bin = PduUtils.typeToBin("float64", js_obj.gps_lat, 8);
        allocator.add(bin, parent_off + 232);
    }
    
    // member: gps_lon, type: float64 (primitive)

    
    {
        const bin = PduUtils.typeToBin("float64", js_obj.gps_lon, 8);
        allocator.add(bin, parent_off + 240);
    }
    
}
